
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Mobile sequencing and analysis in real-time">
      
      
      
      
        <link rel="prev" href="../basic/">
      
      
        <link rel="next" href="../output_files/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Understanding the Dashboard - Nomadic</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/custom.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#read-mapping-statistics" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Nomadic" class="md-header__button md-logo" aria-label="Nomadic" data-md-component="logo">
      
  <img src="../img/little_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nomadic
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Understanding the Dashboard
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-hidden="true"  type="radio" name="__palette" id="__palette_0">
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/JasonAHendry/nomadic.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Nomadic" class="md-nav__button md-logo" aria-label="Nomadic" data-md-component="logo">
      
  <img src="../img/little_logo.svg" alt="logo">

    </a>
    Nomadic
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/JasonAHendry/nomadic.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Basic Usage
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Understanding the Dashboard
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Understanding the Dashboard
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#read-mapping-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Read Mapping Statistics
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#region-coverage-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Region Coverage Statistics
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Region Coverage Statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#identifying-candidate-hrp23-deletions-with-nomads-mvp" class="md-nav__link">
    <span class="md-ellipsis">
      
        Identifying candidate hrp2/3 deletions with NOMADS-MVP
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preliminary-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preliminary Variant Calling
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../output_files/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Output Files & Sharing Data
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Advanced Usage
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    FAQ
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#read-mapping-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Read Mapping Statistics
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#region-coverage-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Region Coverage Statistics
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Region Coverage Statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#identifying-candidate-hrp23-deletions-with-nomads-mvp" class="md-nav__link">
    <span class="md-ellipsis">
      
        Identifying candidate hrp2/3 deletions with NOMADS-MVP
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preliminary-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preliminary Variant Calling
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Understanding the Dashboard</h1>

<p>The <em>Nomadic</em> dashboard provides information about:</p>
<ul>
<li>Read mapping statistics</li>
<li>Region (i.e. amplicon) coverage statistics</li>
<li>Variant calling</li>
</ul>
<!-- - Region (i.e.) amplicon coverage profiles -->

<p>Each of these outputs has it's own panel in the dashboard, as is described in more detail below.</p>
<h2 id="read-mapping-statistics">Read Mapping Statistics</h2>
<p>The top panel in the dashboard <strong>shows information about read mapping</strong>. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Read mapping is the process whereby sequencing reads are aligned to a reference genome based on sequence similarity. The objective of read mapping is to find the location in the reference genome from which a sequencing read likely originated. Sometimes, a single good alignment can be found for a read, but sometimes no good alignments, or multiple equally good alignments are found.</p>
</div>
<p><img alt="read_mapping" src="../img/understand/read_mapping.png" /></p>
<p>In <em>Nomadic</em>, information about read mapping is displayed in a stacked barplot. For each sequenced sample (x-axis), the total height of the bar is the number of alignments produced after all the reads from that sample were mapped. The colours in the stacked bar indicate the type of alignment that was generated for each read:</p>
<ul>
<li>Reads with <strong>only one unique alignment</strong> to the reference genome are shown in dark blue. These are called primary alignments, and are labeled <code>Primary</code>. These are the best for downstream analysis.</li>
<li>Reads with <strong>no alignments</strong> to the reference genome are shown in grey. These are unmapped reads, and labeled <code>Unmapped</code>. This can happen if the read is too short or low quality to be accurately mapped, or derives from an organism other than the reference genome. With NOMADS-MVP, the unmapped reads typically represent offtarget amplification of the host genome.</li>
<li>Reads with <strong>two or more alignments</strong> to the reference are labeled in the lighest shade of blue. These are called secondary alignments, are are labeled <code>Secondary</code>. These can occur when a sequencing read originates from a repetitive region of the genome.</li>
<li>Reads that must be <strong>split in pieces to be aligned</strong> to the reference genome are shown in the middle shade of blue. These are called chimeric alignments, and are labeled <code>Supplementary</code>. These are usually caused by technical factors during library preparation of sequencing, but can also indicate structural varation.</li>
</ul>
<p>The pie chart at the left shows the overall read mapping distribution across all sequenced samples.</p>
<p>Other points to note:</p>
<ul>
<li>In the y-axis labels, 'k' refers to thousands. So 10k = 10,000.</li>
<li>If you are sequencing with NOMADS-MVP, lower parasitemia samples will have more unmapped reads (grey).</li>
</ul>
<h2 id="region-coverage-statistics">Region Coverage Statistics</h2>
<p>The second panel of the dashboard <strong>shows information about the amount of coverage over your target regions</strong>. For amplicon sequencing, the target regions are defined by your amplicons. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Coverage is a measure of the number of sequencing reads that 'cover' a given position of the genome. By position, we mean a single nucleotide in the genome. A read covers that position if, after <em>read mapping</em>, its alignment includes the position. The <em>mean coverage</em> over a region is the mean of the coverage of all the positions included in the region.</p>
</div>
<p><img alt="region_coverage" src="../img/understand/region_coverage.png" /></p>
<p><em>Nomadic</em> shows this information as a scatter plot. For each sample (x-axis), a point will be drawn for each target region. The colour indicates the identity of the target region and there is a legend at the top of the plot. For example, in NOMADS-MVP, the pink points correspond to <em>kelch13</em>. The y-axis gives the mean coverage (<code>mean_cov</code>), number of reads (<code>n_reads</code>) or percentage of bases with greater than 100x coverage (<code>cov_gr100_per</code>), depending on what is selected in the drop down menu located above the pie chart.</p>
<p>The pie chart aggregates information across all samples. In an ideal scenario, the slices of the pie will be close to equal, indicating that the different target regions have similar coverage. In practice, coverage can vary between targets due to differences in PCR efficiency. As long as the lowest abundance amplicon is still recieving sufficient coverage (typically 100x), this is not a major problem.</p>
<h3 id="identifying-candidate-hrp23-deletions-with-nomads-mvp">Identifying candidate <em>hrp2/3</em> deletions with NOMADS-MVP</h3>
<p><img alt="read_coverage_hrp23" src="../img/understand/read_coverage-hrp23-labelled.png" /></p>
<p>You can use the 'Region Coverage Statistics' panel to identify candidate <em>hrp2</em> and <em>hrp3</em> deletions.</p>
<h2 id="preliminary-variant-calling">Preliminary Variant Calling</h2>
<p>The bottom panel of the dashboard <strong>shows information about variant calling</strong>. </p>
<p><img alt="variant_calling" src="../img/understand/variant_calling-ama1.png" /></p>
<p><em>Nomadic</em> shows variant calling information as a heatmap for each target region. You can select different target regions using the drop down menu at the top left.
In the heatmap, the samples are on the x-axis and the variants (i.e. mutations) are on the y-axis. The colour in each cell of the heatmap gives the within-sample alternative allele frequency (WSAF).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The within-sample alternative allele frequency (WSAF) is the <em>fraction</em> of sequencing reads that carry the alternative allele (i.e. variant or mutation). It ranges from 0 to 1.</p>
</div>
<table>
<thead>
<tr>
<th>Colour</th>
<th>WSAF</th>
<th>Interpretation</th>
<th>Conclusion</th>
</tr>
</thead>
<tbody>
<tr>
<td>Dark purple</td>
<td>0.0</td>
<td>None of the reads in the sample carry the mutation.</td>
<td>The sample <strong>does not</strong> have the mutation. It is homozygous reference.</td>
</tr>
<tr>
<td>Dark red</td>
<td>1.0</td>
<td>All of the reads in the sample carry the mutation.</td>
<td>The sample <strong>does</strong> carry the mutation. It is homozygous alternative.</td>
</tr>
<tr>
<td>Light blue, yellow, orange</td>
<td>0.0 &lt; WSAF &lt; 1.0</td>
<td>The WSAF is between 0 and 1. These means some, but not all, of the reads in the sample carry the mutation. For <em>P. falciparum</em> malaria, this indicates that the sample is polyclonal, and at least one clone has the mutation.</td>
<td>The sample <strong>does</strong> carry the mutation. It is heterozygous alternative.</td>
</tr>
<tr>
<td>White</td>
<td>Not available</td>
<td>This indicates that not enough sequencing reads have been collected for the sample to make a determination.</td>
<td>Not enough data. Sample has failed or more sequencing time is required.</td>
</tr>
</tbody>
</table>
<p>In the plot above shows the target <em>ama1</em> for a set of control samples. The first three barcodes are 3D7, HB3 and Dd2. They are monoclonal, so all mutations are either dark purple (WSAF=0) or dark red (WSAF=1). After that, we have mixed strains together at different proportions, making them polyclonal, which leads to intermediate colours. The last three barcodes (46, 47, 48) are negative controls, hence they are the heatmap is white.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.sections", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>