
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Mobile sequencing and analysis in real-time">
      
      
      
      
        <link rel="prev" href="../understand/">
      
      
        <link rel="next" href="../advanced/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Output Files & Sharing Data - Nomadic</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/custom.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#output-files" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Nomadic" class="md-header__button md-logo" aria-label="Nomadic" data-md-component="logo">
      
  <img src="../img/little_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nomadic
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output Files & Sharing Data
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-hidden="true"  type="radio" name="__palette" id="__palette_0">
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/JasonAHendry/nomadic.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Nomadic" class="md-nav__button md-logo" aria-label="Nomadic" data-md-component="logo">
      
  <img src="../img/little_logo.svg" alt="logo">

    </a>
    Nomadic
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/JasonAHendry/nomadic.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Basic Usage
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../understand/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Understanding the Dashboard
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Output Files & Sharing Data
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Output Files & Sharing Data
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output files
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#summaryread_mappingcsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.read_mapping.csv
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summaryregion_coveragecsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.region_coverage.csv
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summaryvariantscsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.variants.csv
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sharing-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Sharing data
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Advanced Usage
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    FAQ
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output files
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#summaryread_mappingcsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.read_mapping.csv
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summaryregion_coveragecsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.region_coverage.csv
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summaryvariantscsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        summary.variants.csv
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sharing-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Sharing data
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Output Files & Sharing Data</h1>

<h2 id="output-files">Output files</h2>
<p><em>Nomadic</em> produces several files that provide information about the quality of your sequencing run, and the variants that were detected. They can be found in the results directory (<code>results/&lt;expt_name&gt;</code>), and are described briefly below:</p>
<h3 id="summaryread_mappingcsv"><code>summary.read_mapping.csv</code></h3>
<p>The <code>summary.read_mapping.csv</code> contains information about read mapping for each sample.</p>
<p>Each row corresponds to a sample, which can be identified by its barcode (e.g. <code>barcode01</code>). The values provided come from <a href="https://www.htslib.org/doc/samtools-flagstat.html"><code>samtools flagstats</code></a> and are counts of read mapping flags inside of the sample's BAM file. Together they summarise how well your sequencing reads mapped to your reference genome.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>barcode</code></td>
<td>Sample barcode.</td>
</tr>
<tr>
<td><code>sample_id</code></td>
<td>Sample ID.</td>
</tr>
<tr>
<td><code>n_total</code></td>
<td>Total number of read alignments.</td>
</tr>
<tr>
<td><code>n_mapped</code></td>
<td>Number of mapped read alignments.</td>
</tr>
<tr>
<td><code>n_unmapped</code></td>
<td>Number of unmapped reads.</td>
</tr>
<tr>
<td><code>n_primary</code></td>
<td>Number of reads mapping uniquely.</td>
</tr>
<tr>
<td><code>n_secondary</code></td>
<td>Number of reads mapping to more than one location.</td>
</tr>
<tr>
<td><code>n_supplementary</code></td>
<td>Number of reads mapping as chimeras.</td>
</tr>
</tbody>
</table>
<p>For more information about read mapping, please see <a href="../understand/#read-mapping-statistics">Understanding the Dashboard</a>.</p>
<h3 id="summaryregion_coveragecsv"><code>summary.region_coverage.csv</code></h3>
<p>The <code>summary.region_coverage.csv</code> file contains information about sequencing coverage over each amplicon in each sample.</p>
<p>Each row contains information about coverage over a specific amplicon in a specific sample. The sample is indicated by its barcode (e.g. <code>barcode01</code>) and the amplicon is indicated by its <code>name</code> (as well as position and length). The values provided come from <a href="https://www.htslib.org/doc/samtools-bedcov.html"><code>samtools bedcov</code></a>.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>barcode</code></td>
<td>Sample barcode.</td>
</tr>
<tr>
<td><code>sample_id</code></td>
<td>Sample ID.</td>
</tr>
<tr>
<td><code>chrom</code></td>
<td>Chromosome of amplicon.</td>
</tr>
<tr>
<td><code>start</code></td>
<td>Start position of amplicon.</td>
</tr>
<tr>
<td><code>end</code></td>
<td>End position of amplicon.</td>
</tr>
<tr>
<td><code>length</code></td>
<td>Length of amplicon.</td>
</tr>
<tr>
<td><code>name</code></td>
<td>Name of amplicon. This comes from the fourth column of BED file used when running <code>nomadic realtime</code> (e.g. <code>-b</code> flag).</td>
</tr>
<tr>
<td><code>n_reads</code></td>
<td>Number of reads mapping to the amplicon.</td>
</tr>
<tr>
<td><code>mean_cov</code></td>
<td>Mean coverage over the amplicon.</td>
</tr>
<tr>
<td><code>cov_gr100</code></td>
<td>Number of positions within the amplicon having &gt;100x coverage.</td>
</tr>
<tr>
<td><code>per_cov_gr100</code></td>
<td>Percentage of positions within the amplicon having &gt;100x coverage.</td>
</tr>
<tr>
<td><code>total_cov</code></td>
<td>Total coverage over amplicon, i.e. mean coverage times length</td>
</tr>
</tbody>
</table>
<p>For more information about region coverage, please see <a href="../understand/#region-coverage-statistics">Understanding the Dashboard</a>.</p>
<h3 id="summaryvariantscsv"><code>summary.variants.csv</code></h3>
<p>The <code>summary.variants.csv</code>contains preliminary information about the variants identified in each sample.</p>
<p>Each row contains information about the genotype, depth, quality, and within-sample allele frequency (WSAF) of a specific single-nucleotide polymorphism (SNP) in a specific sample. For all samples the same set of SNPs are described. The set of SNPs described includes all SNPs where <em>at least one</em> sample carried the alternative allele. Note this file is only generated when the <code>nomadic realtime ... --call</code> flag is used, which is the default.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>barcode</code></td>
<td>Sample barcode.</td>
</tr>
<tr>
<td><code>sample_id</code></td>
<td>Sample ID.</td>
</tr>
<tr>
<td><code>chrom</code></td>
<td>Chromosome of SNP.</td>
</tr>
<tr>
<td><code>pos</code></td>
<td>Position of SNP.</td>
</tr>
<tr>
<td><code>ref</code></td>
<td>Reference nucleotide for SNP.</td>
</tr>
<tr>
<td><code>alt</code></td>
<td>Alternative nucleotide for SNP.</td>
</tr>
<tr>
<td><code>qual</code></td>
<td>Variant quality score of SNP.</td>
</tr>
<tr>
<td><code>mut_type</code></td>
<td>Type of mutation caused by SNP, e.g. synonymous or non-synonymous.</td>
</tr>
<tr>
<td><code>aa_change</code></td>
<td>Amino acid change caused by SNP. For synonymous mutations, we still report (e.g. <code>V380V</code>).</td>
</tr>
<tr>
<td><code>aa_pos</code></td>
<td>Amino acid number containing the SNP.</td>
</tr>
<tr>
<td><code>strand</code></td>
<td>Strand of gene containing the SNP.</td>
</tr>
<tr>
<td><code>amplicon</code></td>
<td>Name of amplicon containing SNP.  This comes from the fourth column of the BED file used when running <code>nomadic realtime</code> (e.g. <code>-b</code> flag).</td>
</tr>
<tr>
<td><code>gt</code></td>
<td>Called SNP genotype for the sample. Can be reference (<code>0/0</code>), heterozygous (<code>0/1</code>), homozygous (<code>1/1</code>) or failed QC (<code>./.</code>). <strong>Note: these are from <code>bcftools call</code> and assume a diploid genome.</strong></td>
</tr>
<tr>
<td><code>gq</code></td>
<td>SNP genotype quality for the sample. Note this is different than variant quality (<code>qual</code>) as it refers to the quality of the genotype call, rather than whether or not the site is variable.</td>
</tr>
<tr>
<td><code>dp</code></td>
<td>Sequencing depth over the SNP. Equivalent to coverage.</td>
</tr>
<tr>
<td><code>wsaf</code></td>
<td>Within-sample alternative allele frequency (<code>ad_alt / (ad_alt + ad_ref)</code>), where <code>ad_ref</code> and <code>ad_alt</code> are the depths of the reference and alternative allele.</td>
</tr>
</tbody>
</table>
<p>For more information about variant calling, please see <a href="../understand/#preliminary-variant-calling">Understanding the Dashboard</a>.</p>
<h2 id="sharing-data">Sharing data</h2>
<p>The results folder for an experiment (<code>results/&lt;expt_name&gt;</code>) contains all the outputs from <em>Nomadic</em>. </p>
<p>Only the CSV files starting <code>summary</code> and the <code>metadata</code> folder are required to relaunch the dashboard with the <code>nomadic dashboard</code> command (see <a href="../basic/">Basic Usage</a>).</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.sections", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>